File: automation-test-playwright/pages/core/CrudListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { OptionalActionHelper } from "@helpers-browser/OptionalActionHelper";
import { BasePage } from "./BasePage";

export class CrudListPage extends BasePage {
  protected readonly path?: string;
  readonly tableBody: Locator;
  readonly filterForm: Locator;
  readonly searchButton: Locator;
  readonly resetButton: Locator;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.tableBody = page.locator("tbody");
    this.filterForm = this.firstVisible(page.locator("form"));
    this.searchButton = page.locator("form button[type='submit'], .btn-filter.btn-search");
    this.resetButton = page.locator(".btn-filter.btn-reset, form button[type='reset']");
  }

  /**
   * Opens the page.
   */
  async open(): Promise<void> {
    if (!this.path) {
      throw new Error("This list page does not define a path.");
    }

    await this.visit(this.path);
  }

  async search(): Promise<void> {
    const button = this.firstVisible(this.searchButton);
    await this.waitForVisible(button);
    await this.waitForEnabled(button);
    await button.click();
  }

  async searchIfAvailable(): Promise<boolean> {
    if (!(await this.searchButton.count())) {
      return false;
    }

    const button = this.firstVisible(this.searchButton);
    await this.waitForVisible(button);
    await this.waitForEnabled(button);
    await button.click();
    return true;
  }

  /**
   * Resets the active filters.
   */
  async resetFilters(): Promise<void> {
    if (await this.resetButton.count()) {
      await this.firstVisible(this.resetButton).click();
    }
  }

  async fillFilter(fieldName: string, value: string): Promise<void> {
    await this.inputByName(fieldName).fill(value);
  }

  async fillFilterIfPresent(fieldName: string, value: string): Promise<boolean> {
    return OptionalActionHelper.fillIfPresent(this.inputByName(fieldName), value);
  }

  async selectFilter(fieldName: string, value: string): Promise<void> {
    await this.inputByName(fieldName).selectOption(value);
  }

  async selectFilterIfPresent(fieldName: string, value: string): Promise<boolean> {
    return OptionalActionHelper.selectIfPresent(this.inputByName(fieldName), value);
  }

  rowByText(text: string): Locator {
    return this.firstVisible(this.page.locator("tbody tr", { hasText: text }));
  }

  firstRowLink(hrefPart: string): Locator {
    return this.firstVisible(this.page.locator(`tbody a[href*="${hrefPart}"]`));
  }

  firstViewButton(): Locator {
    return this.actionButton(this.tableBody, "view");
  }

  firstEditButton(): Locator {
    return this.actionButton(this.tableBody, "edit");
  }

  async clickRowLink(rowText: string, hrefPart: string): Promise<void> {
    await this.firstVisible(this.rowByText(rowText).locator(`a[href*="${hrefPart}"]`)).click();
  }

  async clickFirstRowLink(hrefPart: string): Promise<void> {
    await this.firstRowLink(hrefPart).click();
  }

  async clickFirstViewButton(): Promise<void> {
    await this.firstViewButton().click();
  }

  async clickFirstEditButton(): Promise<void> {
    await this.firstEditButton().click();
  }

  /**
   * Deletes row through the UI.
   */
  async deleteRow(rowText: string): Promise<void> {
    await this.actionButton(this.rowByText(rowText), "delete").click();
  }

  async waitForRowVisible(text: string): Promise<void> {
    await this.waitForVisible(this.rowByText(text));
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/core/CrudListPage.java
package org.suntower.pages.core;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.core.BasePage;
import org.suntower.helpers.browser.OptionalActionHelper;

public class CrudListPage extends BasePage {
  protected String path;
  protected final By tableBody = css("tbody");
  protected final By filterForm = css("form");
  protected final By searchButton = css("form button[type='submit'], .btn-filter.btn-search");
  protected final By resetButton = css(".btn-filter.btn-reset, form button[type='reset']");

  public CrudListPage(WebDriver driver) {
    super(driver);
  }

  public void open() {
    if (path == null || path.isBlank()) {
      throw new IllegalStateException("This list page does not define a path.");
    }
    visit(path);
  }

  public void search() {
    waitForVisible(searchButton);
    waitForEnabled(searchButton);
    click(searchButton);
  }

  public boolean searchIfAvailable() {
    if (!isVisible(searchButton)) {
      return false;
    }
    search();
    return true;
  }

  public void resetFilters() {
    OptionalActionHelper.clickIfPresent(driver, resetButton);
  }

  public void fillFilter(String fieldName, String value) {
    fill(inputByName(fieldName), value);
  }

  public boolean fillFilterIfPresent(String fieldName, String value) {
    return OptionalActionHelper.fillIfPresent(driver, inputByName(fieldName), value);
  }

  public void selectFilter(String fieldName, String value) {
    selectByValue(inputByName(fieldName), value);
  }

  public boolean selectFilterIfPresent(String fieldName, String value) {
    return OptionalActionHelper.selectIfPresent(driver, inputByName(fieldName), value);
  }

  public WebElement rowByText(String text) {
    By rows = css("tbody tr");
    waitForCondition(
        () -> driver.findElements(rows).stream().anyMatch(row -> row.isDisplayed() && row.getText().contains(text)),
        "No table row contained text: " + text);
    return driver.findElements(rows).stream().filter(row -> row.isDisplayed() && row.getText().contains(text)).findFirst().orElseThrow();
  }

  public WebElement firstRowLink(String hrefPart) {
    By link = css("tbody a[href*='" + hrefPart + "']");
    return firstVisible(link);
  }

  public WebElement firstViewButton() {
    return firstVisible(actionButton(tableBody, "view"));
  }

  public WebElement firstEditButton() {
    return firstVisible(actionButton(tableBody, "edit"));
  }

  public void clickRowLink(String rowText, String hrefPart) {
    WebElement row = rowByText(rowText);
    click(row.findElement(By.cssSelector("a[href*='" + hrefPart + "']")));
  }

  public void clickFirstRowLink(String hrefPart) {
    firstRowLink(hrefPart).click();
  }

  public void clickFirstViewButton() {
    firstViewButton().click();
  }

  public void clickFirstEditButton() {
    firstEditButton().click();
  }

  public void deleteRow(String rowText) {
    click(rowByText(rowText).findElement(actionButton("delete")));
  }

  public void waitForRowVisible(String text) {
    rowByText(text);
  }
}
